### ISSUE TITLE
stalls after terminating a foreground job with background processes running

### ISSUE THREAD

**Issue by olivedae** [heapsters/shelldon#3]
Example of this with mpsh and default input/output:

```
mpsh> /usr/local/bin/bomb &
[1] (6875) /usr/local/bin/bomb &
mpsh> /usr/local/bin/bomb 
# SIGINT interruption sent
Job [2] terminated by signal 2
mpsh> jobs 
# waits for job 1 to complete 
Boom!
# no jobs left now, jobs command returns nothing
mpsh>     
```

**Comment by olivedae** [comment]
Seems as if after terminating and reaping the foreground process, our main process becomes stuck in the SIGCHILD handler loop waiting to reap our background processes.

```
# mpsh with logging on
mpsh>  /usr/local/bin/bomb &
# Normal processing for background thread 
[1] (7206) /usr/local/bin/bomb &
mpsh> /usr/local/bin/bomb
EVAL [0]
EVAL [1]
EVAL [2]
EVAL [4]
EVAL [5]
EVAL [5a]
EVAL [5b]
EVAL [6]
WAITFG [0]
WAITFG [1]
WAITFG [2]
EVAL [3]
# SIGINT interruption sent 
TERM [0]
TERM [1]
Job [2] (7027) terminated by signal 2
TERM [2]
WAITFG [3]
mpsh> jobs
REAP [0]
REAP [1]
# background job thread still asleep for several more seconds
Boom!
REAP [1]
REAP [2]
# evaluation of `jobs` command
EVAL [0]
EVAL [1]
# no jobs left now, returns nothing
mpsh>
```

**Comment by olivedae** [comment]
Fixed in ed103cbbd327b5fee51b132d4731c4fb3aea028c. Replaces `wait` with `waitpid` and passing the option flag of WNOHANG.
